### ISSUE TITLE
Vmagent: high CPU and RAM usage with stream aggregation

### ISSUE THREAD

**Issue by hollden** [VictoriaMetrics/VictoriaMetrics#5898]
### Is your question request related to a specific component?

vmagent v1.97.1

### Describe the question in detail

Hi! I am trying to configure stream aggregation at vmagent and send this metrics to another vm storage. But after applying this config, I observe a huge CPU and RAM consumption. 
Here is the graphics:
<img width="712" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/42714896/64704fa3-a670-41dd-a24b-ef138f5b094c">
<img width="704" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/42714896/8f128fe2-96b1-4399-b527-3359d26a1519">
This gaps you can observe, is only because I've set memory limit to 15Gb and it is killed because of OOM constantly.  

Also I noticed, that after enabling streaming aggregation, number of unique samples were increased several times
<img width="701" alt="изображение" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/42714896/10a4e5c7-afc6-4ed6-93ce-8fc006f20404">

My command line flags:
```
 - -promscrape.config=/config/scrape.yml
 - -remoteWrite.tmpDataPath=/tmpData
 - -remoteWrite.url=https://URL-to-cold-data
 - -remoteWrite.url=https://URL-to-hot-data
 - -envflag.enable=true
 - -envflag.prefix=VM_
 - -loggerFormat=json
 - -remoteWrite.streamAggr.config=/config/stream.yml,
 - -remoteWrite.tlsInsecureSkipVerify=true,true
```
and stream config:
```yaml
- match: '{__name__=~".*"}'
  interval: 1m
  outputs: [avg]
  output_relabel_configs:
  - source_labels: [__name__]
    target_label: __name__
    regex: "(.+):.+"
```

Thank you in advance!

### Troubleshooting docs

- [ ] General - https://docs.victoriametrics.com/Troubleshooting.html
- [X] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [ ] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by hagen1778** [comment]
Hello @hollden! We held some work on optimizing resource usage of vmagent during stream aggregation in this PR https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5850.
I'd appreciate if you can give it a try and help us with testing. The docker image for linux amd64 can be found [here](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5850). If you still observe severe resource degradation, please report the following:
1. Ingestion rate in samples/s
2. Approx amount of unique time series processed by vmagent on specified aggregation interval
3. Absolute values of resource usage (CPU, RAM)
4. [Profiles](https://docs.victoriametrics.com/vmagent/#profiling) for CPU and mem.

Thank you!

**Comment by valyala** [comment]
@hollden , the provided [stream aggregation config](https://docs.victoriametrics.com/stream-aggregation/#stream-aggregation-config) calculates the average value over one minute individually per every unique [time series](https://docs.victoriametrics.com/keyconcepts/#time-series). `vmagent` maintains the sum of samples and the number of encountered samples per each unique time series, so it could calculate the average value at the end of every configured interval with the formula: `avg = sum / count`. `vmagent` uses [time series name plus all its labels](https://docs.victoriametrics.com/keyconcepts/#time-series) as a dictionary key for fast lookup of the `(sum, count)` state. If the ingested time series have many long labels (like in typical Kubernetes monitoring case), then `vmagent` may require a lot of RAM for storing all the metric names and labels across all the time series.

There are the following solutions exist, which may help reducing RAM and CPU usage at streaming aggregation:

- To use more specific `match` filter, so the aggregation is applied only to the needed time series.
- To reduce the number of output time series, which are generated by streaming aggregation. This can be done by [aggregating across multiple time series](https://docs.victoriametrics.com/stream-aggregation/#reducing-the-number-of-stored-series).
- To drop long labels from the input time series via `input_relabel_configs` [relabeling](https://docs.victoriametrics.com/stream-aggregation/#relabeling).

As @hagen1778 already said, the https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5850 must significantly reduce memory usage for typical stream aggregation. But it may increase CPU usage by up to 30% at the same time, since `vmagent` spends additional CPU time on compressing long labels.

@hollden , could you give more details about your use case? Why do you need calculating per-minute average across all the input [time series](https://docs.victoriametrics.com/keyconcepts/#time-series)?

P.S. All the changes from https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5850 have been included in the commit 28a9e92b5e4208dfac924f283e946b2fb163a213 , so you can build `vmagent` from this commit according to [these instructions](https://docs.victoriametrics.com/vmagent/#how-to-build-from-sources) and verify whether it uses less RAM on your workload.

P.P.S. The commit 28a9e92b5e4208dfac924f283e946b2fb163a213 adds support `keep_metric_names` option, which can replace the `output_relabel_configs` in your example to:

```yaml
- match: '{__name__=~".*"}'
  interval: 1m
  outputs: [avg]
  keep_metric_names: true
```

**Comment by hollden** [comment]
@valyala @hagen1778 Hi! Thank you for your help!
@valyala I've built vmagent from the commit you mentioned, and tested it with stream config: 
```yaml
  - interval: 1m
  outputs: [avg]
  keep_metric_names: true
```
And it helped!
Now I see reducing memory consumption by 2 times, and thanks for your explanation I understand why vmagent consumes so much of cpu and memory.

Here is the graphics from monitoring dashboards just in case you want to look at them
<img width="702" alt="изображение" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/42714896/344899cf-3320-4547-9e30-ccf0eb187107">
<img width="630" alt="изображение" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/42714896/0e78a94e-4855-46f8-952f-c66d7e710442">
<img width="633" alt="изображение" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/42714896/370abf29-5dfc-450b-80ef-ca07a2e201d6">

Run all metrics through stream config is needed, because I wanted to make 2 storages - with hot and cold data
